action-server/covidflow/actions/test_navigation_form.py

```diff
@@ -0,0 +1,245 @@
+    TestingLocation,
+    TestingLocationPhone,
+    get_testing_locations,
+)
+
+from .constants import LANGUAGE_SLOT, TEST_NAVIGATION_SUCCESS_SLOT
+from .lib.log_util import bind_logger
+
+logger = structlog.get_logger()
+
+FORM_NAME = "test_navigation_form"
+
+
+POSTAL_CODE_REGEX = re.compile(r"[a-z]\d[a-z] ?\d[a-z]\d", re.IGNORECASE)
+MAX_POSTAL_CODE_TRIES = 2
+
+POSTAL_CODE_SLOT = "test_navigation__postal_code"
+INVALID_POSTAL_CODE_COUNTER_SLOT = "test_navigation__invalid_postal_code_counter"
+TRY_DIFFERENT_ADDRESS_SLOT = "test_navigation__try_different_address"
+LOCATIONS_SLOT = "test_navigation__locations"
+END_FORM_SLOT = "test_navigation__end_form"
+
+
+class TestNavigationForm(FormAction):
+    def __init__(self):
+        self.geocoding_client = Geocoding()
+
+        super().__init__()
+
+    def name(self) -> Text:
+        return FORM_NAME
+
+    async def run(
+        self, dispatcher, tracker, domain,
+    ):
+        bind_logger(tracker)
+        return await super().run(dispatcher, tracker, domain)
+
+    @staticmethod
+    def required_slots(tracker: Tracker) -> List[Text]:
+        if tracker.get_slot(END_FORM_SLOT) is True:
+            return []
+
+        if tracker.get_slot(LOCATIONS_SLOT):
+            return [POSTAL_CODE_SLOT]
+
+        return [POSTAL_CODE_SLOT, TRY_DIFFERENT_ADDRESS_SLOT]
+
+    def slot_mappings(self) -> Dict[Text, Union[Dict, List[Dict]]]:
+        return {
+            POSTAL_CODE_SLOT: self.from_text(),
+            TRY_DIFFERENT_ADDRESS_SLOT: [
+                self.from_intent(intent="affirm", value=True),
+                self.from_intent(intent="deny", value=False),
+            ],
+        }
+
+    async def validate_test_navigation__postal_code(
+        self,
+        value: Text,
+        dispatcher: CollectingDispatcher,
+        tracker: Tracker,
+        domain: Dict[Text, Any],
+    ) -> Dict[Text, Any]:
+        postal_code = _get_postal_code(value)
+        if postal_code is None:
+            return _check_postal_code_error_counter(tracker, dispatcher)
+
+        try:
+            geocode = self.geocoding_client.get(postal_code=postal_code)
+            if geocode is None:
+                return _check_postal_code_error_counter(tracker, dispatcher)
+
+            testing_locations = await get_testing_locations(geocode[0], geocode[1])
+
+        except:
+            dispatcher.utter_message(
+                template="utter_test_navigation__could_not_fetch_1"
+            )
+            dispatcher.utter_message(
+                template="utter_test_navigation__could_not_fetch_2"
+            )
+            return {POSTAL_CODE_SLOT: postal_code, END_FORM_SLOT: True}
+
+        if len(testing_locations) == 0:
+            dispatcher.utter_message(template="utter_test_navigation__no_locations")
+            return {
+                POSTAL_CODE_SLOT: postal_code,
+                LOCATIONS_SLOT: [location.raw_data for location in testing_locations],
```

Comment:
if `testing_locations` is empty, why doing this?

Reply:
actually, is there any way of just _if/elsing _ and having a single return at the end? this logic for the return is repeated 3 times

Reply:
Right. That was copy-pasting I didn't think to correct.